akka-stream/src/main/scala/akka/stream/impl/io/SynchronousFilePublisher.scala

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.io
+
+import java.io.{ File, RandomAccessFile }
+import java.nio.channels.FileChannel
+
+import akka.actor.{ ActorLogging, DeadLetterSuppression, Props }
+import akka.io.DirectByteBufferPool
+import akka.stream.actor.ActorPublisherMessage
+import akka.util.ByteString
+
+import scala.annotation.tailrec
+import scala.concurrent.Promise
+
+/** INTERNAL API */
+private[akka] object SynchronousFilePublisher {
+  def props(f: File, completionPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int) = {
+    require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+    require(initialBuffer > 0, s"initialBuffer must be > 0 (was $initialBuffer)")
+    require(maxBuffer >= initialBuffer, s"maxBuffer must be >= initialBuffer (was $maxBuffer)")
+
+    Props(classOf[SynchronousFilePublisher], f, completionPromise, chunkSize, initialBuffer, maxBuffer)
+  }
+
+  final case object Continue extends DeadLetterSuppression
+}
+
+/** INTERNAL API */
+private[akka] class SynchronousFilePublisher(f: File, bytesReadPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int)
+  extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  import SynchronousFilePublisher._
+
+  val buffs = new DirectByteBufferPool(chunkSize, maxBuffer)
+  var eofReachedAtOffset = Long.MinValue
+
+  var readBytesTotal = 0L
+  var availableChunks: Vector[ByteString] = Vector.empty
```

Comment:
Back from the car-repair-shop and investigated the options we're discussing with adding them to the "drain the file" benchmark:

```
[info] FileSourcesBenchmark.fileChannel                   2048  avgt  100  1217.977 ±  61.751  ms/op
[info] FileSourcesBenchmark.fileChannel_noReadAhead       2048  avgt  100  1921.258 ± 116.078  ms/op
[info] FileSourcesBenchmark.fileChannel_stage       2048  avgt   40  2864.094 ± 450.808  ms/op
```

So the naive/ugly readahead does help in those scenarios, the Stage is not very stellar at this point in time (no fusing in action here). 

Let me try to wrap this up (and assuming realistic time bounds):
- we do not have fine control over fusing / not fusing yet. The win by using a stage would be fusing. Translating this guy into a Stage is not very hard if we want to do it later on I believe.
- making the "read ahead as buffer in front of it" works when we fuse, otherwise it's just another async boundary (does not help with perf, I measured that).
- If we want to do the "read ahead and collapse into 1 signal", I need to extend the `ActorPublisher` with `onNext(x, demandDrained = 10)`. I'm not sure we want to expose this as such... It is indeed something more fancy that the AP can do but the Stage not so much. I'm not convienced it's worth it for the 1st impl of this anyway though... 

Summing up:
- if we want a Stage at some point, it's a 15 minute task to change to it from the Actor.
- current best perf for "drain dat file" is the naive readahead (and I think it could be improved) 
- At least the Actor isolates the IO from the rest of the world nicely, and theres the default file-io dispatcher now.

So: how about we go with the simplest impl for now and revisit the topic once fusing is more real? If we don't like the readahead's current form I'm OK with removing it, I'd keep it as an Actor though - sounds good?

----

On a related note, this is what people are currently doing:

```
Source(() ⇒ scala.io.Source.fromFile(file).getLines())
// FileSourcesBenchmark.naive_ioSourceLinesIterator  avgt   20  7067.944 ± 1341.847  ms/op
```

So whichever impl already is very noticable improvement :wink: